### ISSUE TITLE
All threads are blocking at publish event to the buffer

### ISSUE THREAD

**Issue by seanguo** [LMAX-Exchange/disruptor#164]
We have monitored a strange behavior when using log4j2.  Most of the threads are blocked at the following:
> "xxx-t-116" #806381 daemon prio=5 os_prio=0 tid=0x00007fcf55037800 nid=0x200 runnable [0x00007fcebc6e0000]
    java.lang.Thread.State: TIMED_WAITING (parking)
   at sun.misc.Unsafe.park(Native Method)
   at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:338)
   at com.lmax.disruptor.MultiProducerSequencer.next(MultiProducerSequencer.java:136)
   at com.lmax.disruptor.MultiProducerSequencer.next(MultiProducerSequencer.java:105)
   at com.lmax.disruptor.RingBuffer.publishEvent(RingBuffer.java:568)
   at org.apache.logging.log4j.core.async.AsyncLogger.logWithVarargTranslator(AsyncLogger.java:298)
   at org.apache.logging.log4j.core.async.AsyncLogger.logInBackground(AsyncLogger.java:193)
   at org.apache.logging.log4j.core.async.AsyncLogger.logMessageInBackgroundThread(AsyncLogger.java:171)
   at org.apache.logging.log4j.core.async.AsyncLogger.logMessage(AsyncLogger.java:130)
   at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:1011)
   at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:988)
   at org.apache.logging.log4j.spi.AbstractLogger.debug(AbstractLogger.java:234)

But seems the BatchEventProcessor is not busy at that time:

>  "Log4j2-Log4j2Scheduled-2" #18 daemon prio=5 os_prio=0 tid=0x00007fcf547e8800 nid=0x520a waiting on condition [0x00007fcf22e05000]
    java.lang.Thread.State: TIMED_WAITING (parking)
   at sun.misc.Unsafe.park(Native Method)
   - parking to wait for  <0x0000000640043c40> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
   at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
   at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
   at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
   at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
   at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   at java.lang.Thread.run(Thread.java:745)

 > "Log4j2-AsyncLogger[AsyncContext@18b4aac2]1" #17 daemon prio=5 os_prio=0 tid=0x00007fcf54703800 nid=0x5209 runnable [0x00007fcf2350c000]
    java.lang.Thread.State: RUNNABLE
   at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:151)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   at java.lang.Thread.run(Thread.java:745)

We are using disruptor 3.3.4 and I checked the source code in github that line(151) has nothing, it's just a }.  I also noticed the compiled class has some difference when viewed in the decompiler:

 public void run()
   {
/* 108 */     if (!this.running.compareAndSet(false, true))
     {
/* 110 */       throw new IllegalStateException("Thread is already running");
     }
/* 112 */     this.sequenceBarrier.clearAlert();
 
/* 114 */     notifyStart();
 
/* 116 */     Object event = null;
/* 117 */     long nextSequence = this.sequence.get() + 1L;
     try
     {
       while (true)
       {
/* 124 */         long availableSequence = this.sequenceBarrier.waitFor(nextSequence);
 
/* 126 */         while (nextSequence <= availableSequence)
         {
/* 128 */           event = this.dataProvider.get(nextSequence);
/* 129 */           this.eventHandler.onEvent(event, nextSequence, nextSequence == availableSequence);
/* 130 */           nextSequence += 1L;
         }
 
/* 133 */         this.sequence.set(availableSequence);
       }
     }
     catch (TimeoutException e) {
       while (true) notifyTimeout(this.sequence.get());
     }
     catch (AlertException ex)
     {
/* 141 */       while (this.running.get());
     }
     catch (Throwable ex)
     {
       while (true)
       {
/* 148 */         this.exceptionHandler.handleEventException(ex, nextSequence, event);
/* 149 */         this.sequence.set(nextSequence);
/* 150 */         nextSequence += 1L;
       }
 
     }
     finally
     {
/* 156 */       notifyShutdown();
/* 157 */       this.running.set(false);
     }
   }
Any idea about that? Thanks.

**Comment by mikeb01** [comment]
Duplicate of issue #138.

**Comment by seanguo** [comment]
@mikeb01 We are using the sleep strategy. Will the fix for issue #138 also address this issue?

**Comment by mikeb01** [comment]
Reopening as I didn't realise that you were using the SleepingWaitStrategy.

When it fails again, could you post the full stack trace.

**Comment by mikeb01** [comment]
How are you publishing to the ring buffer?

**Comment by seanguo** [comment]
@mikeb01 We are using log4j2 2.5 which uses the disruptor. We do not operate on the ring buffer directly. Just found this maybe a disruptor issue based on the thread dump so I created here.

**Comment by mikeb01** [comment]
I'm confident that this is not a disruptor issue.  It may be related to the default exception handler issue that was fixed recently in Log4J.

**Comment by seanguo** [comment]
@mikeb01  Thanks. Do you mean this ticket https://issues.apache.org/jira/browse/LOG4J2-1324 ?

**Comment by mikeb01** [comment]
Yes.

**Comment by cmuramoto** [comment]
Hi, I believe I've found a similar behavior in production, in just 1 out of 7 servers, after upgrading it to JDK1.7.0_76.

I'm using RingBuffer with WorkHandler, essentially in a Many (producer) to One (consumer). The number of consumers is configured to be the same as the number of CPUs.

```java
public EventProcessor(){
//this implements EventFactory, ThreadFactory, ExceptionHandler, WorkHandler
disruptor = new Disruptor<>(this, cap, this, ProducerType.MULTI, new SleepingWaitStrategy());
//
disruptor.setDefaultExceptionHandler(this);
disruptor.handleEventsWithWorkerPool(createWorkers());
}

//creates as many workers as available CPUs
private WorkHandler<EventHolder>[] createWorkers() {
		final WorkHandler[] rv = new WorkHandler[Runtime.getRuntime().availableProcessors()];
		for (int i = 0; i < rv.length; i++) {
			rv[i] = this;
		}

		return rv;
}

//prevent Worker Threads from dying
public void handleEventException(final Throwable ex, final long sequence, final EventHolder event) {
		log.warn(ex.getMessage(), ex);
}
```

Stack dump shows several ajp Threads (which are used by the load balancer) in a similar state:

```
ajp-172.16.100.5-8009-1 tid=355 [WAITING] [DAEMON]
java.util.concurrent.locks.LockSupport.parkNanos(long) LockSupport.java:349
com.lmax.disruptor.MultiProducerSequencer.next(int) MultiProducerSequencer.java:136
com.lmax.disruptor.MultiProducerSequencer.next() MultiProducerSequencer.java:105
com.lmax.disruptor.RingBuffer.publishEvent(EventTranslatorTwoArg, Object, Object) RingBuffer.java:509
---<omitted>---.java:130
---<omitted>---.java:110
---<omitted>---.java:466
---<omitted>---:355
---<omitted>---:173
org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(ServletRequest, ServletResponse) FilterChainProxy.java:342
---<omitted>---:62
org.springframework.security.web.FilterChainProxy.doFilter(ServletRequest, ServletResponse, FilterChain) FilterChainProxy.java:160
---<omitted>---:37
org.jboss.web.tomcat.security.SecurityAssociationValve.invoke(Request, Response) SecurityAssociationValve.java:182
org.terracotta.modules.tomcat.tomcat_5_5.SessionValve55.tcInvoke(Request, Response) SessionValve55.java:100
org.terracotta.modules.tomcat.tomcat_5_5.SessionValve55.invoke(Request, Response) SessionValve55.java:85
org.terracotta.session.ModernTomcatSessionValve.invoke(Request, Response) ModernTomcatSessionValve.java:66
java.lang.Thread.run() Thread.java:745
```

I collected 2 heap dumps, with a 20min gap between them, and every sequence (cursor,gatingSequenceCache, sequences of gatingSequences) in the **MultiProducerSequencer** were exactly the same. RingBuffer's entries were all empty (My events hold 2 states, which are set by EventTranslatorTwoArg and set to null after processing).

All the servers host the same application and the only difference among them that presented the issue runs with JDK 1.7.0_76. Other servers run with JDK 1.7.0_60. 

I don't know if there's some kind of JVM-JIT related issue, but after setting -XX:-UseLoopPredicate on the problematic server, the issue was gone (well, it didn't present itself so far in about a week).

**Comment by mikeb01** [comment]
Please can you raise a separate issue?  The problem you're reporting seems to be unrelated to Log4J.

**Comment by mikeb01** [comment]
I believe that this is a Log4J problem, with the management of creating/destroy ring buffers and error handling.
